### PR TITLE
Upgrade Tensorflow to 2.7

### PR DESCRIPTION
**Pull request by abelfodil**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [ ] I've linked relevant issues in the PR body
- [ ] I've applied [the relevant labels](https://intranet.neuro.polymtl.ca/geek-tips/contributing#pr-labels-a-href-pr-labels-id-pr-labels-a) to this PR
- [ ] I've applied a [release milestone](https://github.com/spinalcordtoolbox/spinalcordtoolbox/milestones) (major, minor, patch) in line with [Semantic Versioning guidelines](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Misc%3A-Creating-a-new-release#convention-for-naming-releases) 
- [ ] I've assigned a reviewer

#### PR contents

- [ ] I've consulted [SCT's internal developer documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Programming%3A-Tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Programming%3A-Documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

- Builds upon https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3646 (don't review the file changes just yet)
- Uses Keras that is shipped with TF2
- Upgrade to TF 2.7.0 (I just put the latest version but I can change it to an earlier one)
- Update documentation for hard-core pythonistas
- I'm waiting for the pipeline to run to validate the changes (installation worked tho)

### CODE REVIEW

File: dev/install/install_dependencies_with_workarounds.py [hunk 1]
```diff
@@ -1,35 +1,11 @@
...
 # coding: utf-8
-# Check if the platform requires conda tensorflow.
+# Check if the platform requires conda skimage
 #
 # Author: Christian S. Perone
 
-import platform, subprocess
+import subprocess
 
-def tf_workarounds():
```
Comment (abelfodil) [inline dev/install/install_dependencies_with_workarounds.py:9]: not sure if we should remove

File: install_sct [hunk 2]
```diff
@@ -625,46 +625,14 @@ python -m pip install -U "pip!=21.2.*"
...
   # We use "--ignore-installed" to preserve the version of `certifi` installed into the conda
   # env, which prevents https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3609
-  pip install -r "$REQUIREMENTS_FILE" --ignore-installed certifi &&
   print info "Installing spinalcordtoolbox..." &&
-  pip install -e .
+  pip install -e . --ignore-installed certifi -f https://download.pytorch.org/whl/cpu/torch_stable.html
 ) ||
   die "Failed running pip install: $?"
 
-## Check if CPU supports AVX instruction set, and install precompiled TensorFlow if it isn't,
```
Comment (abelfodil) [inline install_sct:646]: not sure if we should remove

File: dev/install/install_dependencies_with_workarounds.py (hunk 1 above)
Comment (joshuacwnewton) [reply dev/install/install_dependencies_with_workarounds.py:9]: Ah, so. The `dev/` folder is largely... dead code? There are a lot of things in that folder that were once relevant, but are no longer touched. SCT hasn't exactly had a great procedure for pruning bits that are no longer needed?

So, as far as changes that affect present-day SCT, you can safely ignore almost everything in `/dev/`

File: install_sct (hunk 2 above)
Comment (joshuacwnewton) [reply install_sct:646]: This will need to be updated (rather than removed). Though, I'm fairly certain that we'll be able to find (or compile ourselves) new `.whl` files for whichever version of TF we upgrade, though. There are lots of people out there who have an interest in providing support for AVX-less CPUs. :)

For now, you can just ignore this? I think it'll be easiest to just focus on getting TF working for our CI runners. And then, once we can confirm that it's possible to upgrade, we can go back and fill in the gaps for AVX-less users.

File: setup.py [hunk 3]
```diff
@@ -52,7 +48,9 @@ def get_dependencies(requirements_path=None):
     'scipy',
     'scikit-image',
     'scikit-learn',
-    'tensorflow~=1.15.0',
+    'intel-tensorflow~=2.7.0; sys_platform != "darwin"',
```
Comment (abelfodil) [inline setup.py:51]: Fix for https://github.com/onnx/onnx-tensorflow/issues/535

**Comment by abelfodil** [comment]
If the observable behavior differs from what we want, we could call https://www.tensorflow.org/api_docs/python/tf/compat/v1/disable_v2_behavior

File: setup.py [hunk 4]
```diff
@@ -52,7 +48,9 @@ def get_dependencies(requirements_path=None):
...
+    'tensorflow~=2.7.0; sys_platform == "darwin"',
```
Comment (abelfodil) [inline setup.py:52]: @joshuacwnewton do you happen to know if there's a MacOS TF2 version with MKL support? Right now the vanilla CPU version can't run ConvXD layers correctly :(

I got around this issue for Windows and Linux by using `intel-tensorflow`. Unfortunately I don't have a Mac machine that I can legally use for this purpose.

Comment (abelfodil) [reply setup.py:52]: Current `tensorflow-metal` package requires Python 3.{8,9} so using it is out of the question :( Looking for alternatives for `tensorflow` accelerations

Comment (joshuacwnewton) [reply setup.py]: One small note:: The only thing keeping us from upgrading to Python 3.8/3.9 _is_ TensorFlow (2.2 and 2.5 respectively). (https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3367)

So, upgrading both at the same time wouldn't be out of the question!

Comment (joshuacwnewton) [reply setup.py]: One other comment I want to make: In SCT, we're merely using TF/Keras to perform inference using pre-trained models (as opposed to training models from scratch). So, we actually don't have much of a need for GPU-acceleration? In fact, it may even be preferable to stick to CPU-only libraries, since that would typically reduce the installation size of SCT.

So, if you have a working CPU-based upgrade to TF, then I would say that's more than good enough. :)

Comment (joshuacwnewton) [reply setup.py]: Here's my attempt at summarizing the exploration you've done (just to make sure I'm understanding OK):

---

Upgrading stock `tensorflow` to 2.X causes a error to be thrown re: `NHWC` vs. `NCHW` data formatting. Specifically:

```
CPU implementation of Conv3D currently only supports the NHWC tensor format.
```

(Ah! I'm just realizing I came across the exact same error in my previous investigations, too: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3367#issuecomment-907357925)

This is the root issue, and there are [three possible workarounds](https://github.com/onnx/onnx-tensorflow/issues/535#issuecomment-568743917) for this issue (according to [the linked comment](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3653#discussion_r787271777), anyway):

  1. Using an MKL-based version of Tensorflow:
      * [`intel-tensorflow`](https://pypi.org/project/intel-tensorflow/) is a pip package that provides MKL support, but only for Linux/Windows, and not macOS. 
      * [`tensorflow-mkl`](https://anaconda.org/anaconda/tensorflow-mkl) is a conda package that provides MKL support for macOS, but only for TF 2.0.0. (Dead-end for macOS support.)
  2. Using a GPU-based version of Tensorflow:
       * [`tensorflow-metal`](https://pypi.org/project/tensorflow-metal/) is a pip package that provides GPU support for macOS, but only for Python 3.8/3.9. (And it would also [involve PluggableDevices](https://pypi.org/project/tensorflow-metal/), which may require more complicated prerequisites on the user side of things.)
   3. Trying to... modify the data format on the model side of things?
       * As far as I can tell, we haven't explored this option yet, right?

Comment (joshuacwnewton) [reply setup.py]: There's a potential "workaround 4." in [this comment](https://github.com/onnx/onnx-tensorflow/issues/535#issuecomment-683366977):

> setting `os.environ["CUDA_VISIBLE_DEVICES"]="-1"` (as mentioned in this reply [#614 (comment)](https://github.com/onnx/onnx-tensorflow/issues/614#issuecomment-625251872)) fixes the issue for me.

That linked reply (from onnx-tensorflow issue 614) seems pretty illuminating to me! I didn't really understand _why_ the NWHC/NCHW formatting issues were a problem in the first place, but that explanation maybe clarifies things a little? :eyes:

Comment (abelfodil) [reply setup.py]: That's a good summary of the situation. For "workaround 4", I don't think it works because `tensorflow-cpu` package fails as well.

**Comment by joshuacwnewton** [comment]
Closing this PR, as it's been superseded by #3738, which removes Tensorflow entirely from SCT.

**Comment by abelfodil** [comment]
That's great!
